zeebe/atomix/utils/src/test/java/io/atomix/utils/concurrent/SingleThreadContextTest.java

```diff
@@ -49,4 +54,26 @@ public void shouldInvokeHandlerOnException() throws InterruptedException {
     // then
     assertEquals(0, latch.getCount());
   }
+
+  @Test
+  public void shouldRetryCallableUntilSuccessful() throws InterruptedException {
+    final var success = new AtomicBoolean(false);
+    final var delayStrategy =
+        new ExponentialBackoffRetryDelay(Duration.ofMillis(10), Duration.ofMillis(1));
+    final var result =
+        threadContext.retryUntilSuccessful(
+            () -> {
+              if (!success.get()) {
+                throw new RuntimeException("Expected");
+              }
+            },
+            delayStrategy);
+
+    Thread.sleep(250);
```

Comment:
:wrench: I would advise against thread sleep, as this can easily be flaky. Typically we use things like CyclicBarrier or CountDownLatch to control asynchronous execution. Another option would be using an `AtomicInteger` in the retry runnable, use awaitility to wait until it's been called a couple times, and then flip the switch, and then assert eventually it completes :+1: